plugin/storage/badger/factory.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2018 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package badger
+
+import (
+	"expvar"
+	"flag"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/dgraph-io/badger"
+	"github.com/spf13/viper"
+	"github.com/uber/jaeger-lib/metrics"
+	"go.uber.org/zap"
+
+	depStore "github.com/jaegertracing/jaeger/plugin/storage/badger/dependencystore"
+	badgerStore "github.com/jaegertracing/jaeger/plugin/storage/badger/spanstore"
+	"github.com/jaegertracing/jaeger/storage/dependencystore"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+var (
+	// ValueLogSpaceAvailable returns the amount of space left on the value log mount point in bytes
+	ValueLogSpaceAvailable = expvar.NewInt("badger_value_log_bytes_available")
+	// KeyLogSpaceAvailable returns the amount of space left on the key log mount point in bytes
+	KeyLogSpaceAvailable = expvar.NewInt("badger_key_log_bytes_available")
+	// LastMaintenanceRun stores the timestamp of the previous maintenanceRun
+	LastMaintenanceRun = expvar.NewInt("badger_storage_maintenance_last_run")
+	// LastValueLogCleaned stores the timestamp of the previous ValueLogGC run
+	LastValueLogCleaned = expvar.NewInt("badger_storage_valueloggc_last_run")
```

Comment:
badger itself exposes metrics (which are not replicated here) using expvar. If these are not exposed by default, then can I somehow encapsulate expvar available metrics through MetricsFactory?

Reply:
Prometheus does have expvar Collector, but I didn't see it being called as an option in the MetricsFactory. So should it be implemented there or are they on purpose ignored by default?

Reply:
I think we're talking about two different things here. The above code declares metrics emitted by the Jaeger code. Everywhere in Jaeger we're using MetricsFactory for that, and the user can choose which metrics backend they want, including expvar or prometheus. So there's no strong reason to hardcode just the expvar implementation for those.

Separately, you're saying that Badger internals are exposing expvar metrics too. This one is slightly more difficult, because iirc we don't currently enable expvar http handler if the metrics backend is not expvar. Maybe we should always enable the /debug/expvar endpoint even if prometheus is the main metrics selection. Alternatively, if prom client has an adapter that can expose expvar via prom scraping endpoint, we could use that.